genkit-tools/telemetry-server/src/localFileTraceStore.ts

```diff
@@ -141,3 +226,115 @@ export class LocalFileTraceStore implements TraceStore {
   }
 }
+
+export interface IndexSearchResult {
+  stopIndex?: number;
+  data: Record<string, string>[];
+}
+
+/**
+ * A super basic searchable index implementation. It's not particularly efficient,
+ * but should not be worse than reading individual trace files from disk.
+ */
+export class Index {
+  private currentIndexFile: string;
+
+  constructor(private indexRoot: string) {
+    // TODO: do something about index getting too big. Delete/forget old stuff, etc.
+    this.currentIndexFile = path.resolve(this.indexRoot, '1.idx');
+    fs.mkdirSync(this.indexRoot, { recursive: true });
+    if (!fs.existsSync(this.currentIndexFile)) {
+      fs.writeFileSync(this.currentIndexFile, '');
+    }
+  }
+
+  add(traceData: TraceData) {
+    const rootSpans = Object.values(traceData.spans).filter(
+      (s) => !s.parentSpanId
+    );
+    const rootSpan = rootSpans.length > 0 ? rootSpans[0] : undefined;
+    const indexData = {
+      id: traceData.traceId,
+    } as Record<string, string | number>;
+    if (rootSpan?.attributes?.['genkit:type']) {
+      indexData['type'] = `${rootSpan.attributes['genkit:type']}`;
+    }
+    if (rootSpan?.startTime) {
+      indexData['start'] = rootSpan.startTime;
+    }
+    if (rootSpan?.displayName) {
+      indexData['name'] = rootSpan.displayName;
+    }
+    if (rootSpan?.endTime) {
+      indexData['end'] = rootSpan.endTime;
+    }
+
+    try {
+      lockfile.lockSync(lockFile(this.currentIndexFile));
+
+      fs.appendFileSync(
+        this.currentIndexFile,
+        JSON.stringify(indexData) + '\n'
+      );
+    } finally {
+      lockfile.unlockSync(lockFile(this.currentIndexFile));
+    }
+  }
+
+  search(query: {
+    limit: number;
+    startFromIndex?: number;
+    filter?: TraceQueryFilter;
+  }): IndexSearchResult {
+    const idxTxt = fs.readFileSync(this.currentIndexFile, 'utf8');
+    const fullData = idxTxt
+      .split('\n')
+      .map((l) => {
+        try {
+          return JSON.parse(l);
+        } catch {
+          return undefined;
+        }
+      })
+      .filter((d) => {
+        if (!d) return false;
+        if (!query?.filter) return true;
+        if (
+          query.filter.eq &&
+          Object.keys(query.filter.eq).find(
+            (k) => d[k] !== query.filter!.eq![k]
+          )
+        ) {
+          return false;
+        }
+        if (
+          query.filter.neq &&
+          Object.keys(query.filter.neq).find(
+            (k) => d[k] === query.filter!.neq![k]
+          )
+        ) {
+          return false;
+        }
+        return true;
+      })
+      .reverse();
+
+    // do pagination
+    const startFromIndex = query.startFromIndex ?? 0;
+
+    const result = {
+      data: fullData.slice(startFromIndex, startFromIndex + query.limit),
+    } as IndexSearchResult;
+
+    // if there are more results, populate stop index.
+    if (startFromIndex + query.limit < fullData.length) {
+      result.stopIndex = startFromIndex + query.limit;
```

Comment:
nit: Say 'paginationToken' or 'paginationLastIndex' or 'lastIndex' instead of 'stopIndex'?

Reply:
renamed to pageLastIndex